### PR TITLE
[PM-5742] Rework Usage of Extension APIs that Cannot be Called with the Background Service Worker

### PR DESCRIPTION
**Pull request by cagonzalezcs**
## Type of change

```
- [ ] Bug fix
- [ ] New feature development
- [x] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

The changes in this PR ensure that we do not run into any conflicts when calling extension APIs within manifest v3. It ensure we replace deprecated or modified API calls [based on the approach recommended by Google's migration guide](https://developer.chrome.com/docs/extensions/develop/migrate/api-calls#replace-unsupported-apis).

It also takes into account API calls that can only be made from a foreground context, ensuring that any calls made from the background script fallback to a default value such that execution of the script is not interrupted. 

These changes should not impact manifest v2 in any significant manner.

## Code changes

- **apps/browser/src/autofill/content/notification-bar.ts:** Replacing the call for `chrome.extension.getURL` to the supported `chrome.runtime.getURL`.
- **apps/browser/src/platform/browser/browser-api.spec.ts:** Incorporating Jest tests for changes made to the `BrowserApi` class.
- **apps/browser/src/platform/browser/browser-api.ts:** Updating the call to `getBackgroundPage` to return a default `null` value if the method is not available, as is the case within a background service worker script. Implementing a `getExtensionViews` method that will return an empty array if called within a background service worker, and will return the expected array of window objects if called in a foreground context.
- **apps/browser/src/platform/services/browser-platform-utils.service.spec.ts:** Adding jest tests to validate changes madw ithin `BrowserPlatformUtilsService`
- **apps/browser/src/platform/services/browser-platform-utils.service.ts:** Updating calls to `chrome.extension.getViews` to use the new `BrowserApi.getExtensionViews` method.
- **apps/browser/test.setup.ts:** Adding mocks for the `extension` browser API.

### CODE REVIEW

File: apps/browser/src/platform/browser/browser-api.ts [hunk 1]
```diff
@@ -199,20 +199,54 @@ export class BrowserApi {
...
+    }
+
+    return chrome.extension.getViews(fetchProperties);
+  }
+
+  /**
+   * Queries all extension views that are of type `popup`
+   * and returns whether any are currently open.
+   */
   static async isPopupOpen(): Promise<boolean> {
```
Comment (MGibson1) [inline apps/browser/src/platform/browser/browser-api.ts:248]: Is this still meaningful? does the foreground ever try to determine if the popup is open or is it just the background?

It feels like this not working in service workers likely causes issues.

**Review by MGibson1 — APPROVED** [review]
All the code looks good, my only concern is that these changes represent lost functionality in mv3 for the service worker. Does it just happen that we aren't using them in the background, or is a fix coming later for these lost apis?

File: apps/browser/src/platform/browser/browser-api.ts (hunk 1 above)
Comment (cagonzalezcs) [reply apps/browser/src/platform/browser/browser-api.ts:248]: We use this within the `SettingsComponent.import` method, so it is used in the foreground...

That said, we also use it within the `BrowserPlatformUtilsService.isViewOpen` method, which is used by the `VaultTimeoutService.checkVaultTimeout` method... 

That method does run in the background, and as a result we'll be running into problems in depending on it to identify if a popup is currently open.

With that said, you're likely right in suggesting we deprecate this method entirely (usage of it is minimal overall) and consider rewriting this check to function as a message to the popup. If the message is responds, then we know an extension view is currently open. If not, then we know it is closed.

**Comment by cagonzalezcs** [comment]
> All the code looks good, my only concern is that these changes represent lost functionality in mv3 for the service worker. Does it just happen that we aren't using them in the background, or is a fix coming later for these lost apis?

My initial reworking of this didn't take into account lost functionality for the mv3 service worker, but your comment has me thinking a bit about that. So this is what I've found thus far.

`BrowserApi.getBackgroundPage` is used in three different locations in the build; within the popup `services.module.ts` file when setting the `mainBackground` variable, within `BrowserPopupUtils.backgroundInitializationRequired` and in the method `BrowserApi.isBackgroundPage`.

We are aware that the usage of `BrowserApi.getBackgroundPage` within `services.module.ts` needs to be reworked in its entirety, as that presents a distinct block in terms of having a functional  vault popup page.

`BrowserPopupUtils.backgroundInitializationRequired` is used within `services.module.ts` as well as `BrowserPopupUtils.inPrivateMode`. The method `BrowserPopupUtils.inPrivateMode` is used within the `private-mode-warning.component.ts` file, which displays a warning to Firefox users. 

As with the usage of `BrowserApi.getBackgroundPage` within `services.module.ts`, we likely will be deprecating the usage of `BrowserPopupUtils.backgroundInitializationRequired` entirely, as the circumstances for how services are loaded into the vault popup page should remove the necessity for this kind of check. 

`BrowserApi.isBackgroundPage` is used within `BrowserApi.addListener` and `BrowserApi.removeListener`. These methods leverage a check for the background page to ensure that we memoize event listeners attached to the `chrome.runtime.onMessage` listener, specifically within Safari extensions when loading them in the vault popup page. We do this to deregister the listeners after usage of them to avoid memory leaks (I'm not sure this is the best approach, but it is what we have in place right now). 

We also use `BrowserApi.isBackgroundPage` within `SessionSyncer.update` and `SessionSyncer.updateSession`. It seems `SessionSyncer.update` leverages this check to handle calling the `memoryStorageService.save` method when we are loaded in manifest v2. `SessionSyncer.updateSession` handles calling `memoryStorageService.save` when we are loaded either in manifest v3 or when the background page is the current `self` value. In either instance, not sure exactly what is going on here, but we likely would need to revisit this with the mv3 implementation. 

`BrowserApi.getExtensionViews` is the other major change, and I discussed what gets affected in a previous comment. The biggest concern I have is with how we handle usage of the `VaultTimeoutService.checkVaultTimeout` method. That is used pretty extensively in the background, and will no longer remain valid now that we cannot pull extension views in this manner. 

We could consider adjusting this to work as a `chrome.tabs.query` call instead, but I'm unsure what limitations exist here when it comes to identifying `sidebar` type views. Would have to experiment a bit.

**Comment by cagonzalezcs** [comment]
Continuing on from the above...

So the incompatibilities for manifest v3 with the `BrowserApi.getExtensionViews` method might be relatively simple to resolve. 

The biggest concern here is with the `BrowserPlatformUtilsService.isViewOpen` method, which is used by the `VaultTimeoutService.checkVaultTimeout` method to determine whether the extension should lock. If the user currently has an extension view open, we avoid locking the extension.

This is something that can be easily checked without using `BrowserApi.getExtensionViews`. If we establish a `chrome.runtime.onMessage` listener strictly within the extension popup (say within the `InitService` used to bootstrap the popup's service classes), we can then ping the listener using a `chrome.runtime.sendMessage` call. if the listener is triggered, it would return a `true` value to indicate "I have an open popoup", and if the listener is not available the `chrome.runtime.sendMessage` call our receive an `undefined` response.

**Comment by MGibson1** [comment]
It actually sounds like for most of these we plan on not using them in mv3 -- session syncer goes away after the state provider updates.

I like that answer quite a lot.

It also sounds like your solutions for open views would work.

File: apps/browser/src/platform/browser/browser-api.ts (hunk 1 above)
Comment (MGibson1) [reply apps/browser/src/platform/browser/browser-api.ts:248]: Created a ticket to reflect this work

**Review by MGibson1 — APPROVED** [review]

**Review by jprusik — APPROVED** [review]

**Pull request merged**